src/main/java/com/o19s/es/termstat/TermStatScorer.java

```diff
@@ -0,0 +1,155 @@
+        StatisticsHelper tf_stats = new StatisticsHelper();
+        StatisticsHelper tp_stats = new StatisticsHelper();
+
+        PostingsEnum postingsEnum = null;
+        for (Term term : this.terms) {
+            if (docID() == DocIdSetIterator.NO_MORE_DOCS) {
+                break;
+            }
+
+            TermStates termStates = TermStates.build(searcher.getTopReaderContext(), term, scoreMode.needsScores());
+
+            assert termStates != null && termStates
+                    .wasBuiltFor(ReaderUtil.getTopLevelContext(context));
+
+            TermState state = termStates.get(context);
+
+            if (state == null) {
+                continue;
+            }
+
+            // Collection Statistics
+            df_stats.add(termStates.docFreq());
+            idf_stats.add(sim.idf(termStates.docFreq(), searcher.getIndexReader().numDocs()));
+            ttf_stats.add(termStates.totalTermFreq());
+
+            // Doc specifics
+            TermsEnum termsEnum = context.reader().terms(term.field()).iterator();
+            termsEnum.seekExact(term.bytes(), state);
+            postingsEnum = termsEnum.postings(postingsEnum, PostingsEnum.ALL);
+
+            // Verify document is in postings
+            if (postingsEnum.advance(docID()) == docID()){
+                tf_stats.add(postingsEnum.freq());
+
+                if(postingsEnum.freq() > 0) {
+                    StatisticsHelper positions = new StatisticsHelper();
+                    for (int i = 0; i < postingsEnum.freq(); i++) {
+                        positions.add((float) postingsEnum.nextPosition() + 1);
+                    }
+                    tp_stats.add(positions.getAggr(posAggr));
+                } else {
+                    tp_stats.add(0.0f);
+                }
+            // If document isn't in postings default to 0 for tf/tp
+            } else {
+                tf_stats.add(0.0f);
+                tp_stats.add(0.0f);
+            }
+        }
+
+        // Prepare computed statistics
+        StatisticsHelper computed = new StatisticsHelper();
+        HashMap<String, Float> termStatDict = new HashMap<>();
+        Bindings bindings = new Bindings(){
+            @Override
+            public DoubleValuesSource getDoubleValuesSource(String name) {
+                return DoubleValuesSource.constant(termStatDict.get(name));
+            }
+        };
+
+        for(int i = 0; i < tf_stats.getSize(); i++) {
+            // Update the term stat dictionary for the current term
+            termStatDict.put("df", df_stats.getData().get(i));
+            termStatDict.put("idf", idf_stats.getData().get(i));
+            termStatDict.put("tf", tf_stats.getData().get(i));
+            termStatDict.put("tp", tp_stats.getData().get(i));
+            termStatDict.put("ttf", ttf_stats.getData().get(i));
+
+            DoubleValuesSource dvSrc = compiledExpression.getDoubleValuesSource(bindings);
```

Comment:
Ideally this should be called once when creating the scorer, this will drastically improve the perf I believe since here you rescan the term enum on every doc.
Might not be trivial to achieve but I think it might worth it (perhaps in a separate subsequent PR?)

Reply:
I've played with this a bit but haven't figured out the magic to get it working yet.